### ISSUE TITLE
Split Client / Server Logic

### ISSUE THREAD

**Issue by mcdonnelldean** [mcollina/hyperemitter#14]
As discussed with Matteo, the client and server logic should be split.

**Comment by mcollina** [comment]
I think we should split the whole P2P logic in its own module, that reuses
this one.
Il giorno ven 22 mag 2015 alle 23:26 Dean McDonnell <
notifications@github.com> ha scritto:

**Comment by mcdonnelldean** [comment]
I started work on this. The only part I got stuck with was that we are polluting the hyperlog with Peer connection messages which will have to be handled by the user.

**Comment by mcdonnelldean** [comment]
Of course this is because my assumption is the peer and replication works via Hyperlog. The peer will manage connection and replication and will provide you with Hyperlog to read or append to.

**Comment by mcollina** [comment]
The key point here is that HyperEmitter uses itself to distribute the peer
information. It's very meta. If you end up doing it with hyperlog only, you
will end up with something like HyperEmitter as it is now.

EventPeer messages are automatically handled by HyperEmitter now. The user
doesn't have to do anything.
Il giorno dom 24 mag 2015 alle 13:37 Dean McDonnell <
notifications@github.com> ha scritto:

**Comment by mcdonnelldean** [comment]
@mcollina Yup, that was what I was getting stuck. So in terms of the logic, what belongs in the peer. Without HyperLog it won't distribute itself. Is it the case that we simply define the infrastructure around connecting to one or more streams and it is up to the consumer of the lib to manage self distribution and peer discovery?

So the new module would:
- Create the server stream and pass it back to the consumer.
- Provide the logic to connect to _n_ Peers but the consumer would have to call (not self called).
- Handle closing and clean up logic.

The end result would be a module that connects to streams, it would be up to the consumer (HyperEmitter in this case) to do something useful with the stream, including sending replication events on Peer connection.

**Comment by mcollina** [comment]
I think something that sits on top of HyperEmitter, which would need to
expose HyperLog replicate method. Basically, taking all the current
listen/connect logic and putting it in another module. So, it will not be
HyperEmitter that knows about HyperPeer, but the inverse.

Does it makes any sense to you?
Il giorno dom 24 mag 2015 alle 14:12 Dean McDonnell <
notifications@github.com> ha scritto:

**Comment by mcdonnelldean** [comment]
Ah!!! The other way around does make much more sense 😊

**Comment by mcdonnelldean** [comment]
As of #15 we are in good shape to support this now.
